src/base/Assert.hh

```diff
@@ -33,6 +35,13 @@
  * Postcondition debug assertion macro. Use to "ensure" that return values or
  * side effects are as expected when leaving a function.
  */
+/*!
+ * \def INSIST
```

Comment:
I am not sure I grasp correctly the spelling[/word choice] here.   What are some example where you would use this macro?

Reply:
The word choice is another holdover from Shift (and LANL's Draco/DS++ before that, see https://github.com/lanl/Draco/blob/develop/src/ds%2B%2B/Assert.hh ). The example case would be:
```c++
std::ifstream critical_setup_data(path);
INSIST(critical_setup_data, "Important file '" << path << "' failed to open");
```

Reply:
Fair enough.  So it seems that this can/should be used for problems that:
  * expected to happens sometimes/rarely
  * can not be recovered
  * are fixable, from outside the program, by the user.
  
In this case, why isn't it using the error logger?

In other word, what is the difference between INSIST and
```
If (!critical_setup_data)
   CELER_LOG(critical) << "Important file '" << path << "' failed to open";
```
?

Reply:
`CELER_LOG(critical)` just shows a log message; `INSIST` throws an exception (which could in theory be caught by the caller code). So in a standalone app (assuming `INSIST` in the library code) I would have:
```c++
try
{
    run();
    return EXIT_SUCCESS;
}
except (const std::runtime_error& e)
{
    CELER_LOG(critical) << e.what();
    return EXIT_FAILURE;
}
```
the INSIST throws a nicely worded error as a string, and then the frontend decides that it should be printed as a critical  error before exiting.

Reply:
humm ... I disagree with that implementation of critical (it would then basically "just" be an error with the word critical in it ... why even have it ...  and the documentation agrees with me :) :
```
28:    critical,   //!< Something went terribly wrong; we're aborting now! Bye!
```
(on the other hand, the code seems to agree with you :) ).

In addition, if the exception is not thrown the result can be less than satisfactory, or so it seems:
```
#include <exception>
#include <string>

int main(int, char**)
{
  throw std::string("some nicely formatted message with all the explanations");
  return 0;
}
```
```
libc++abi.dylib: terminating with uncaught exception of type std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >
Abort trap: 6
```

Reply:
I'll update the logger text to be less ambiguous -- I was trying to convey the perspective of `main` making one function call to `LOG(critical)` to inform the user about the pending program exit, and a second call to return/exit/abort/terminate. I don't think it's sensible for some log "levels" to forcibly exit the code and others to not -- logging should just be how and when to print messages IMHO. 

Yep, uncaught exceptions are ugly, but they're less ugly than aborting immediately or keeping running? Recall also that the testing harness (googletest) has its own exception handlers, so it can catch `INSIST` errors and continue to run tests or at least clean up the test nicely rather than immediately exiting.

Reply:
i.e. ideally the fatal/insist/critical thing would both log a message and throw an exception.

Reply:
I strongly disagree -- loggers should log, not assert or abort. The Python logger only logs; it even has `log.exception` for printing logging information about exceptions. Is there some other logging utility that you use which kills the program at a certain log level?

Reply:
PS This recalls to mind one of the early discussions from the "GeantX pilot project" days -- I think our resolution then was to neither log nor assert and just worry about it later 😆

Reply:
> Is there some other logging utility that you use which kills the program at a certain log level?

The one for ROOT ... but yes it is 20 years old :)

> I strongly disagree -- loggers should log, not assert or abort.

That's fair enough.  Still I insist that INSIST should both log and then throw.

For that matters, I still having not having an easy time reading that:
```
std::ifstream configuration_file(path);
INSIST(configuration_file, "Important file '" << path << "' failed to open");
```
would be always on and throw an exception.  What about
```
std::ifstream configuration_file(path);
if (!configuration_file) FATAL("Important file '" << path << "' failed to open");
or
if (!configuration_file) CRITICAL_ERROR("Important file '" << path << "' failed to open");
```
humm .. I see, it does not let us record/print the condition.
I guess the full name is akin to:
```
REQUIRE_THEN_THROW(configuration_file, "Important file '" << path << "' failed to open");
```

Reply:
Aha, I didn't know ROOT did that. That explains the expectations -- I'm more familiar with python, and you with ROOT.

The way that `CELER_LOG` works (printing the message in the destructor) precludes throwing an exception, no matter how much we want it to :\

`REQUIRE` always throws if the condition isn't met -- this is more like a `REQUIRE_VERBOSE`, except that it doesn't get disabled when `CELERITAS_DEBUG` is false.

Reply:
So here's a demo of using the INSIST inside a google test and a standalone app:
```
[ RUN      ] UnitsTest.equivalence
unknown file: Failure
C++ exception with description "../test/base/Constants.test.cc:33:
celeritas: runtime error: 4 == 5 failed:
    There are four lights" thrown in the test body.
[  FAILED  ] UnitsTest.equivalence (0 ms)
```
and the unit test continues.

For a naive app that doesn't catch std::runtime_error, the default error handler can print the text. (Your earlier example threw a non-`std::exception` object, which is probably why it didn't print meaningful output.)
```
libc++abi.dylib: terminating with uncaught exception of type celeritas::RuntimeError: ../app/demo.cc:9:
celeritas: runtime error: 4 == 5 failed:
    There are four lights
Abort trap: 6
```

So in either case, an unhandled exception gets the text correctly printed even without a `CELER_LOG` call. And if the user code wants to handle the exception, it seems unfair that we should force it to be printed in a certain way beforehand.

Reply:
That output is indeed good enough.

Reply:
Fair enough.  We probably need to find a better way to distinguish.  We have REQUIRE/CHECK/ENSURE/CHECK_UNREACHABLE that are on only in debug builds that are undistinguishable from INSIST that is always on.

Reply:
You're right, the "always on" vs "debug only" distinction for the macros is definitely something that isn't obvious. And the assertion names are really only clear by context (and for me by the years of seeing them in code).

I'm open to ideas on how to make the difference more clear...

Reply:
What about switch from a macro style to a function (naming) style to indicate it does not change with the configuration:
```
std::ifstream configuration_file(path);
CelerThrowIf( !configuration_file, "Important file '" << path << "' failed to open");
```
Or ```CELER_THROW_IF```

Reply:
I'd prefer an obvious macro style for macros; I've been bitten before many times by identifiers that look like macros but aren't, and vice-versa. (See also Microsoft's old macro definitions of min, max, and a bunch of other surprising identifiers.)

I'm OK with `CELER_XXX` as below, but I'm hesitant to make "convenience" macro names so long that they're no longer convenient -- `CELER_PRECONDITION_ASSERT`, `CELER_INTERNAL_ASSERT`, `CELER_THROW_RUNTIME_ERROR_EVEN_WHEN_NOT_DEBUGGING` are all very obvious but very clunky.

I think the most reasonable situation is to have short verbs (prefixed with `CELER_` to avoid conflicts) that aren't completely intuitive but are at least intelligible after having seen one or two use cases. The following are a little long for my taste but are definitely more obvious than the DBC-themed macros:
- `CELER_ASSERT_PRE`: precondition assertions
- `CELER_ASSERT`: general internal assertions
- `CELER_ASSERT_POST`: postcondition assertions
- `CELER_ASSERT_UNREACHABLE`: (should never be hit and helps compiler out if debugging is disabled)
- `CELER_VASSERT_RUNTIME`: verbose assertion for runtime behavior?

Maybe with another iteration or two we could hit on a better set of macro names.

Reply:
- `CELER_PRE_COND`: precondition assertions
- `CELER_ASSERT`: general internal assertions
- `CELER_POST_CON`: postcondition assertions
- `CELER_UNREACHABLE`: (should never be hit and helps compiler out if debugging is disabled)
- `CELER_ERROR`:  handle (foreseeable) unrecoverable runtime error  (or `CELER_FATAL_ERROR`)

Reply:
Maybe drop the second underscore for pre/post? And I don't think `CELER_ERROR(condition, failure_msg)` parses as well as another "assurance"-type verb such as `CELER_INSIST(condition, failure_msg)`. It doesn't convey that it's only an error if the condition is false.

So the differences from `CELER_ASSERT` are:
- Throws a `RuntimeError` instead of `DebugError` (i.e. it's intended for conditions that aren't the programmer's fault)
- Does not get disabled when `CELERITAS_DEBUG` is false
- Has a second argument that is a verbose error message
I'm increasingly skeptical that there's a concise way to convey all three semantic differences -- so I suggest that a completely different word (such as "insist" vs "assert") may be the most elegant solution.

- `CELER_PRECOND`: precondition assertions
- `CELER_ASSERT`: general internal assertions
- `CELER_POSTCOND`: postcondition assertions
- `CELER_UNREACHABLE`: (should never be hit and helps compiler out if debugging is disabled)
- `CELER_INSIST`:  handle (foreseeable) unrecoverable runtime error  (or `CELER_FATAL_ERROR`)

Reply:
I've moved this conversation to an independent discussion at https://github.com/celeritas-project/celeritas/discussions/97.